### PR TITLE
[CALCITE-2535] Runtime failure check not working in SqlTesterImpl.java

### PR DESCRIPTION
**Pull request by zhztheplayer**
This is the [JIRA](https://issues.apache.org/jira/browse/CALCITE-2535).

1. Make run time failure check work as expected;
2. Lazy initial the generated static field in DeterministicCodeOptimizer.java to avoid ExceptionInInitializerError.

### CODE REVIEW

File: linq4j/src/test/java/org/apache/calcite/linq4j/test/DeterministicTest.java [hunk 1]
```diff
@@ -423,15 +467,15 @@ private boolean isConstant(Expression e) {
         equalTo("{\n"
             + "  return new Runnable(){\n"
             + "      int test() {\n"
-            + "        return "
-            + "$L4J$C$java_math_BigInteger_valueOf_42L_add_java_math_BigInteger_valued8d57d69;\n"
+            + "        return ((java.math.BigInteger) (java.math.BigInteger) $L4J$C$java_math_BigInteger_valueOf_42L_.get()).add((java.math.BigInteger) (java.math.BigInteger) $L4J$C$java_math_BigInteger_valueOf_42L_.get());\n"
```
Comment (vlsi) [inline linq4j/src/test/java/org/apache/calcite/linq4j/test/DeterministicTest.java]: Note: previously `BigInteger.valueOf(42).add(BigInteger.valueOf(42))` was memoized, however now you perform `.add` in the runtime.

Could you please keep that optimization in place?

File: core/src/test/java/org/apache/calcite/sql/test/AbstractSqlTester.java [hunk 2]
```diff
@@ -307,15 +317,17 @@ public SqlTester withOperatorTable(SqlOperatorTable operatorTable) {
...
     return with("connectionFactory", connectionFactory);
   }
 
-  protected SqlTesterImpl with(final String name, final Object value) {
-    return new SqlTesterImpl(factory.with(name, value));
+  protected SqlTester with(final String name, final Object value) {
+    return createImpl(factory.with(name, value));
   }
 
+  protected abstract SqlTester createImpl(SqlTestFactory factory);
```
Comment (vlsi) [inline core/src/test/java/org/apache/calcite/sql/test/AbstractSqlTester.java]: Could you please name consistent naming? `SqlTester with(SqlTestFactory factory)`

File: linq4j/src/test/java/org/apache/calcite/linq4j/test/DeterministicTest.java (hunk 1 above)
Comment (zhztheplayer) [reply linq4j/src/test/java/org/apache/calcite/linq4j/test/DeterministicTest.java]: So sorry for this mistake, fixed now.

**Comment by zhztheplayer** [comment]
Closed this as the discussion in JIRA.

**Pull request closed without merging**